README.md
# r4 - A stack-based VM/CPU with human-readable machine language

r4 is an simple, minimal, and interactive environment where the source code IS the machine code. There is no compilation in r4.

## What is r4?

r4 is a stack-based, RPN, virtual CPU/VM that supports many registers, functions, locals, floating point, and any amount of RAM.

A register (a built-in variable) is identified by consecutive UPPERCASE characters. They can be retrieved, set, increment, or decremented in a single operation (r,s,i,d).

r4 converts A-Z into a base 26 number, so A=0, Z=25, BA=26. ZZZ=17575 (26^3-1), and ZZZZ=456975. This value is then used as an index into an array of registers or function vectors, So it is extremely fast, but not very memory-efficient. Modern PCs have enough memory to be able to support 4 char names. A Teensy4 can support 3 char names. A UNO might only be able to handle 1 char names.

Function are defined in a Forth-like style, using ':', and you call them using the 'c' opcode. For example:

- 0(CPY (N F T--): copy N bytes from F to T)
- :CPY s2 s1 0[r1 C@ r2 C! i1 i2];
- 123 rF rT cCPY 0(copy 123 bytes from rF to rT)

The number of registers, functions, and user memory are configurable and can be scaled as necessary to fit into a system of any size. For example, on an ESP8266 board, a typical configuration might be 676 (26*26) registers and functions, and 24K of user RAM. In such a system, the register names would be in the range of [AA..ZZ], and function names would be in the range of [AA..ZZ]. On a Arduino Leonardo, you might configure the system to have 13 registers, 26 functions, and 1K user RAM. On a RPI Pico, you can have 676 registers and functions, with 64K RAM.

- Example 1: "Hello World!" - the standard "hello world" program.
- Example 2: 1 sA 2 sB 3 sC rA rB rC ++ . -would print 6.
- Example 4: 32 126\[13,10,I#." - ",\] - would print the ASCII table
- Example 3: The typical Arduino "blink" program is a one-liner, except this version stops when a key is pressed:

    1000 sS 13 xPO 1{\ 0 1[I 13 xPWD rS xW] K? 0=} K@ \

Examples for r4 are here: https://github.com/CCurl/r4/blob/main/examples.txt

## Why did I create r4?

There are multiple goals for r4:

- Freedom from the need for a multiple gigabyte tool chain and the edit/compile/run/debug loop for developing everyday programs. Of course, you need one of these monsters to build r4, but at least after that, you are free of them.
- Many programming environments use tokens and a large SWITCH statement in a loop to execute the user's program. In those systems, the machine code (aka - byte-code ... the cases in the SWITCH statement) are often arbitrarily assigned and are not human-readable, so they have no meaning to the programmer when looking at the code that is actually being executed. Additionally there is a compiler that is needed in order to work in that environment. In these enviromnents, there is a steep learning curve; the programmer needs to learn: (1) the user environment, (2) the hundreds or thousands of user functions (or "words" in Forth), and (3) how they work together. I wanted to avoid as much as that as possible, and have only one thing to learn: the machine code.
- A simple, minimal, and interactive programming environment that is easy to modify and enhance.
- An environment that can be deployed to many different types of development boards via the Arduino IDE.
- The ability to use the same environment on my personal computer as well as development boards.
- Short commands so that there is not a lot of typing needed.

## The implementation of r4

- The entire system is implemented in a few files, primarily: config.h, r4.h, r4.cpp, pc-main.cpp, and r4.ino.
  - There are a few additional files to support optional functionality (e.g.-WiFi and File access).
- The same code runs on Windows, Linux, and multiple development boards (via the Arduino IDE).
- See the file "config.h" for system configuration settings.

## Memory areas
There are 2 memory areas in r4:
- USER: r4 function code goes in this area. Default size is 128K.
- VARS: use with 'V' opcode. Default size is 256K.
  - The USER area can also be used for data. See the 'U' opcode.
  - The VARS area is not used by the r4 system at all, 100% free.

## Locals

r4 allocates 10 local variables (0-9) per function call. They are referred to, and have the same operations, like registers (r,s,i,d). For example, i4 increments local #4.

## WiFi support

Some boards, for example the ESP8266, support WiFi. For those boards, the __WIFI__ directive can be #defined to enable the boards WiFi.
Note that those boards usually also have watchdogs that need to be enabled via the __WATCHDOG__ #define.

## LittleFS support

Some development boards support LittleFS. For those boards, the __LITTLEFS__ directive can be #defined to save and load the defined code to the board, so that any user-defined words can be reloaded across boots.

## A simple block editor

r4 includes a simple block editor. Many thanks to Alain Theroux for his inspiration.

## Building r4

- The target machine/environment is controlled by the #defines in the file "config.h"
- For Windows, There is a Visual Studio solution. Use the x86 configuration (32-bit).
- For Development boards, I use the Arduino IDE. See the file "config.h" for board-specific settings.
- For Linux systems, there is a makefile. Default is 64-bit, or you can change it to 32-bit.
- I do not have an Apple system, so I haven't tried to build r4 for that environment.
  - However, being such a simple and minimal C program, it should not be difficult to port r4 to any environment.

##  r4 Reference

### INTEGER OPERATIONS
| OP |Stack |Description|
|:-- |:--   |:--|
| +  |  (a b--n)   |n: a+b - addition
| -  |  (a b--n)   |n: a-b - subtraction
| *  |  (a b--n)   |n: a*b - multiplication
| /  |  (a b--q)   |q: a/b - division
| M  |  (a b--r)   |r: a%b - modulo
| S  |  (a b--q r) |q: div(a,b), r: modulo(a,b)  (SLASH-MOD)


### FLOATING POINT OPERATIONS
#### NOTES:
- r4 uses double wide (64-bit) floating point numbers

| OP |Stack |Description|
|:-- |:--   |:--|
| FF | (i--f)    |Convert TOS from integer to float
| FI | (f--i)    |Convert TOS from float to integer
| F+ | (a b--n)  |Float: add
| F- | (a b--n)  |Float: subtract
| F* | (a b--n)  |Float: multiply
| F/ | (a b--n)  |Float: divide
| F< | (a b--f)  |f: if (a < b), else 0
| F= | (a b--f)  |f: if (a = b), else 0
| F> | (a b--f)  |f: if (a > b), else 0
| F. | (n--)     |Float: print top of fload stack


### BIT MANIPULATION OPERATIONS
| OP |Stack |Description|
|:-- |:--   |:--|
| b& | (a b--n)   |n: a and b
| b| | (a b--n)   |n: a or b
| b^ | (a b--n)   |n: a xor b
| b~ | (a--b)     |b: not a      (e.g - 1011 => 0100)
| L  | (a n--b)   |b: a << n     (Left-Shift)
| R  | (a n--b)   |b: a >> n     (Right-Shift)


### STACK OPERATIONS
| OP |Stack |Description|
|:-- |:--   |:--|
| #  | (a--a a)       |Duplicate TOS (DUP)
| \  | (a b--a)       |Drop TOS (DROP)
| $  | (a b--b a)     |Swap top 2 stack items (SWAP)
| %  | (a b--a b a)   |Push 2nd (OVER)
| ~  | (a--b)         |b: -a (NEGATE)
| D  | (a--b)         |b: a-1 (1-)
| P  | (a--b)         |b: a+1 (1+)
| A  | (a--b)         |b: absolute value of a (ABS)
| xK | (--)           |Display the stack (.S)


### MEMORY OPERATIONS
| OP |Stack |Description|
|:-- |:--   |:--|
| @  | (a--n)      |Fetch CELL n from address a
| !  | (n a--)     |Store CELL n  to  address a
| C@ | (a--n)      |Fetch BYTE n from address a
| C! | (n a--)     |Store BYTE n  to  address a
| U  | (n--a)      |a: address of byte n in the USER area.
| V  | (n--a)      |a: address of byte n in the VARS area.


### REGISTERS OPERATIONS
#### NOTES:
- A register reference consecutive UPPERCASE characters.
- The number of registers is controlled by the NUM_REGS #define in "config.h"
- Register A is the same as register AAA, B <-> AAB, Z <-> AAZ
- r"TEST" will push the value of register AAA and then print TEST

| OP |Stack |Description|
|:-- |:--   |:--|
| rABC | (--v)      |v: value of register ABC.
| sABC | (v--)      |v: store v to register ABC.
| iABC | (--)       |Increment register ABC.
| dABC | (--)       |Decrement register ABC.


### LOCALS OPERATIONS
#### NOTES:
- On each function call, 10 locals [r0..r9] are allocated.
- They are de-allocated when ';' executes.
- Locals are NOT initialized.

| OP |Stack |Description|
|:-- |:--   |:--|
| rN | (--v)  |v: value of local #N.|
| sN | (v--)  |v: store v to local #N.|
| iN | (--)   |Increment local N.|
| dN | (--)   |Decrement local n.|


### FUNCTIONS OPERATIONS
#### NOTES:
- A function reference is consecutive UPPERCASE characters.
- The number of functions is controlled by the NUM_FUNCS #define in "config.h"
- Function A is the same as function AAA, B <-> AAB, Z <-> AAZ
- :"TEST"; will define function #0 (A).
- Returning while inside of a loop will eventually cause a problem.
  - Use '^' to unwind the loop stack first.

| OP |Stack |Description|
|:-- |:--   |:--|
| :ABC  | (--)   |Define function ABC. Copy chars to (HERE++) until closing ';'.
| cABC  | (--)   |Call function ABC. Handles "tail call optimization".
| ;     | (--)   |Return: PC = rpop()

 
### INPUT/OUTPUT OPERATIONS
| OP |Stack |Description|
|:-- |:--   |:--|
| .     | (N--)    |Output N as a decimal number
| ,     | (N--)    |Output N as a character (EMIT)
| "str" | (--)     |Output characters until the next '"'
| B     | (--)     |Output a single SPACE (32,)
| N     | (--)     |Output a single NEWLINE (13,10,)
| K?    | (--f)    |f: non-zero if char is ready to be read, else 0.
| K@    | (--n)    |n: Key char, wait if no char is available.
| 0..9  | (--N)    |Scan DECIMAL number N until non digit
|       |          |- to specify multiple values, separate them by space (4711 3333)
|       |          |- to enter a negative number, use "negate" (eg - 490~)
|hXXX   | (--N)    |Scan HEX number N until non hex-digit ([0-9,A-F] only ... NOT [a-f])
| 'C    | (n)      |n: the ASCII value of C
| `x`   | (a--a b) |Copy following chars until closing '`' to (a++).
|       |          |- a: address, b next byte after trailing NULL.


### LOGICAL/CONDITIONS/FLOW CONTROL OPERATIONS
| OP |Stack |Description|
|:-- |:--   |:--|
| <  | (a b--f)    |f: if (a < b) then 1, else 0;
| =  | (a b--f)    |f: if (a = b) then 1, else 0;
| >  | (a b--f)    |f: if (a > b) then 1, else 0;
| _  | (x--f)      |f: if (x = 0) then 1, else 0; (logical NOT)
| (  | (f--)       |if (f != 0), execute code in '()', else skip to matching ')'
| X  | (a--)       |if (a != 0), execute/call function at address a
| G  | (a--)       |if (a != 0), go/jump to function at address a

history.txt
777xBL
fC
fBB
123F<100F<F/F.
:MIL 1000#**;
1 fMIL
500*
xT$1[]xT$-." ms"
100F<11F<F/F.
fC
fF
17 xIC4*xIAF+
\
xIC*xIAF+@
##ee
xIC3*+
\
20 xIC*xIAF+@e
2 xIC*xIAF+@e
xIH.
'/
'\
/-
100sT
rT.
"/"rTxW
"/"rTxW"-"rTxW
"/"rTxW8,"-"rTxW
"/"rTxW8,"-"rTxW8,"\"rTxW
"/"rTxW13,"-"rTxW13,"\"rTxW13,"-"
500sT
"/"rTxW13,"-"rTxW13,"\"rTxW13,"-"
1 10["/"rTxW13,"-"rTxW13,"\"rTxW13,"-"]
1 10[13,"/"rTxW13,"-"rTxW13,"\"rTxW13,"-"]
1 10[13,"/"rTxW13,"-"rTxW13,"\"rTxW13,"-"rTxW]
fHLP
.
999xBL
10xBL
fC
fCSI"?25l"
fCSI"?25h"
1xBL
10xBL
fC
fCOF 1 10[13,"/"rTxW13,"-"rTxW13,"\"rTxW13,"-"rTxW] fCON
200sT
fCOF 1 10[13,"/"rTxW13,"-"rTxW13,"\"rTxW13,"-"rTxW] fCON
100sT fCOF 1 10[13,"/"rTxW13,"-"rTxW13,"\"rTxW13,"-"rTxW] fCON
50sT fCOF 1 10[13,"/"rTxW13,"-"rTxW13,"\"rTxW13,"-"rTxW] fCON
10sT fCOF 1 10[13,"/"rTxW13,"-"rTxW13,"\"rTxW13,"-"rTxW] fCON
25sT fCOF 1 10[13,"/"rTxW13,"-"rTxW13,"\"rTxW13,"-"rTxW] fCON
25sT fCOF 1 20[13,"/"rTxW13,"-"rTxW13,"\"rTxW13,"-"rTxW] fCON
25sT fCOF 1 20[13,rI.B] fCON
100sT fCOF 1 20[13,rI.B] fCON

100sT fCOF 1 20[13,rI.BrTxW] fCON
500sT fCOF 1 20[13,rI.BrTxW] fCON
250sT fCOF 1 20[13,rI.BrTxW] fCON
100sT fCOF 1 20[13,rI.BrTxW] fCON
1000sT fCOF 1 20[13,rI.BrTxW] fCON
100sT fCOF 1 2000[13,rI.BrTxW] fCON
100sT fCOF 1 200[13,rI.BrTxW] fCON
xQ
888xBL
777xBL
eC
eBB
262144#+eCPB
262144#+#+eCPB
262144#+#+#+eCPB
262144#+#+#+#+eCPB
eHLP
.
999xBL
1xBL
eHLP
eC
1 10[" "] 1 1["/"] 1 1["\"]
1 10[1 10[" "] 1 1["/"] 1 1["\"]E]
1 10[1 rI 10-[" "] 1 rI["/"] 1 rI["\"]E]
1 10[1 rI 10-[" "] 1 rI["/"] "|" 1 rI["\"]E]
1 10[1 rI 10-[" "] 1 rI["/"] "|" 1 rI["\"]E]1 21[34,]
1 10[1 rI 10-[" "] 1 rI["/"] "|" 1 rI["\"]E]" "1 21[34,]
1 10[1 rI 10-[" "] 1 rI["/"] "|" 1 rI["\"]E]"  "1 21[34,]
1 10[1 rI 10-[" "] 1 rI["/"] "|" 1 rI["\"]E]"  "1 8[34,]
1 10[1 rI 10-[" "] 1 rI["/"] "|" 1 rI["\"]E]"  "1 8[34,]"|||"
1 10[1 rI 10-[" "] 1 rI["/"] "|" 1 rI["\"]E]"  "1 98[34,]"|||"1 9[34,]
1 10[1 rI 10-[" "] 1 rI["/"] "|" 1 rI["\"]E]"  "1 9[34,]"|||"1 9[34,]
1 10[1 rI 10-[" "] 1 rI["/"] "|" 1 rI["\"]E]"  "1 9[34,]"|||"1 9[34,]1 10[B]"|||"
1 10[1 rI 10-[" "] 1 rI["/"] "|" 1 rI["\"]E]"  "1 9[34,]"|||"1 9[34,]E1 10[B]"|||"
1 10[1 rI 10-[" "] 1 rI["/"] "|" 1 rI["\"]E]"  "1 9[34,]"|||"1 9[34,]E1 11[B]"|||"
1 10[1 rI 10-[" "] 1 rI["/"] "|" 1 rI["\"]E]"  "1 9[34]"###"1 9[34,]E1 11[B]"###"
1 10[1 rI 10-[" "] 1 rI["/"] "|" 1 rI["\"]E]"  "1 9[34,]"|||"1 9[34,]E1 11[B]"|||"
\\\\\\\\\\\\\\\\\\\\\\\\\\\\
1 10[1 rI 10-[" "] 1 rI["/"] "|" 1 rI["\"]E]"  "1 9[34,]"|||"1 9[34,]E1 11[B]"|||"
1 10[1 rI 10-[" "] 1 rI["/"] "|" 1 rI["\"]E]"  "1 9[34,]"###"1 9[34,]E1 11[B]"###"
1 10[1 rI 10-[" "] 1 rI["/"] "|" 1 rI["\"]E]"  "1 9[34,]"|||"1 9[34,]E1 11[B]"|||"
:XM 1 10[1 rI 10-[" "] 1 rI["/"] "|" 1 rI["\"]E]"  "1 9[34,]"|||"1 9[34,]E1 11[B]"|||";
eXM
eC
1 10[1 rI 10-[" "] 1 rI["/"] "|" 1 rI["\"]E]"  "1 9[34,]"|||"1 9[34,]E1 11[B]"|||"
1 20[1 rI 10-[" "] 1 rI["/"] "|" 1 rI["\"]E]"  "1 9[34,]"|||"1 9[34,]E1 11[B]"|||"
1 20[1 rI 20-[" "] 1 rI["/"] "|" 1 rI["\"]E]"  "1 9[34,]"|||"1 9[34,]E1 11[B]"|||"
1 20[1 rI 20-[" "] 1 rI["/"] "|" 1 rI["\"]E]"  "1 19[34,]"|||"1 19[34,]E1 11[B]"|||"
1 20[1 rI 20-[" "] 1 rI["/"] "|" 1 rI["\"]E]"  "1 19[34,]"|||"1 19[34,]E1 22[B]"|||"
1 20[1 rI 20-[" "] 1 rI["/"] "|" 1 rI["\"]E]"  "1 19[34,]"|||"1 19[34,]E1 21[B]"|||"
1 20[1 rI 20-[" "] 1 rI["/"] "|" 1 rI["\"]E]"  "1 19[34,]"|||"1 19[34,]E1 21[B]"|||"
eC
10xBL
eC
eCLL
32 40 eSTC1 10[1 rI 10-[" "] 1 rI["/"] "|" 1 rI["\"]E]"  "1 9[34,]"|||"1 9[34,]E1 11[B]"|||"eCLC
32 40 eSTC 1 10[1 rI 10-[" "] 1 rI["/"] "|" 1 rI["\"]E]"  "1 9[34,]"|||"1 9[34,]E1 11[B]"|||"eCLC
10xBL
eCLU
eCLL
93 100 eSTC eCLS
eC
eCLC
eCLS
32 40 eSTC 1 10[1 rI 10-[" "] 1 rI["/"] "|" 1 rI["\"]E]"  "1 9[34,]"|||"1 9[34,]E1 11[B]"|||"eCLC
100xr
.
100xr.
100xr sX 50 xr sY 7xr sC
eCLLL
eCLL
100xr sX 50 xr sY 7xr 1+ sC
100xr sX 50 xr sY 7xr 31+ sC
rX rY eC
\\
100xr sX 50 xr sY 7xr sC
rX rY eGXY rC 30 eSTC "*" 1000 xW eCLC
eCLC
eC
eCLS
eC
eCLL
37 40 eSTC
xQ
10 xBL
eCLL
77 4 eGXY"*" 1000xW8,
77 4 eGXY"*" 1000xW8,
77 4 eGXY"*" 1000xW8," "
100xr 4 eGXY"*" 1000xW8," "
100xr 40xr eGXY"*" 1000xW8," "
100xr 40xr eGXY"*" 1000xW8," "
100xr 40xr 7 xr 31+ 40 eSTC eGXY"*" 1000xW8," "
100xr 40xr 7 xr 31+ 40 eSTC eGXY"*" 1000xW8," " eCLC
100xr 40xr 7 xr 31+ 40 eSTC eGXY"*" 1000xW8," " eCLC
1 100 [100xr 40xr 7 xr 31+ 40 eSTC eGXY"*" 100xW8," " eCLC]
eCOF eCLS 1 100 [100xr 40xr 7 xr 31+ 40 eSTC eGXY"*" 100xW8," " eCLC]eCON
eC
1xBL
10xBL
eCOF eCLS 1 100 [100xr 40xr 7 xr 31+ 40 eSTC eGXY"*" 100xW8," " eCLC]eCON
eCOF eCLS 1 10 [100xr 40xr 7 xr 31+ 40 eSTC eGXY"*" 250xW8," " eCLC]eCON
eCOF eCLS 1 10 [100xr 40xr 7 xr 31+ 40 eSTC eGXY"*" 500xW8," " eCLC]eCON
eCOF eCLS 1 10 [100xr 40xr 7 xr 30+ 40 eSTC eGXY"*" 500xW eCLC]eCON
eCOF eCLS 1 1000 [100xr 40xr 7 xr 30+ 40 eSTC eGXY"*" 50xW eCLC]eCON
eCOF eCLS 1 10000 [100xr 40xr 7 xr 30+ 40 eSTC eGXY"*" eCLC]eCON
eCOF eCLS 1 1000 [100xr 40xr 7 xr 30+ 40 eSTC eGXY"*" 10xW eCLC]eCON
eCOF eCLS 1 10000 [100xr 40xr 7 xr 30+ 40 eSTC eGXY"*" 10xW eCLC]eCON
eCOF eCLS 1 1000000 [100xr 40xr 7 xr 30+ 40 eSTC eGXY"*" 10xW eCLC]eCON
eCOF eCLS 1 10000 [100xr 40xr 8 xr 30+ 40 eSTC eGXY"*" 10xW eCLC]eCON
\\\\\\\\\\\\\\\\\\\\\
10xBL
eCOF eCLS 1 10000 [100xr 40xr 8 xr 30+ 40 eSTC eGXY"*" 10xW eCLC]eCON
eCOF 1 10000 [100xr 40xr 8 xr 30+ 40 eSTC eGXY"*" 10xW eCLC]eCON

r4.cpp
// R4 - A Minimal Interpreter

#include "r4.h"
#include "config.h"

byte ir, isBye = 0, isError = 0;
addr pc, HERE;
CELL n1, t1, seed = 0;
int dsp, rsp, lsp, locStart;
ST_T   dstack[STK_SZ+1];
CELL   reg[NUM_REGS], locals[(RSTK_SZ+1)*10], lstack[LSTACK_SZ+1];
addr   rstack[RSTK_SZ+1], func[NUM_FUNCS];
byte   user[USER_SZ], vars[VARS_SZ];

void push(CELL v) { if (dsp < STK_SZ) { dstack[++dsp].i = v; } }
CELL pop() { return (dsp) ? dstack[dsp--].i : 0; }

void rpush(addr v) { if (rsp < RSTK_SZ) { rstack[++rsp] = v; } }
addr rpop() { return (rsp) ? rstack[rsp--] : 0; }

void vmInit() {
    dsp = rsp = lsp = locStart = 0;
    for (int i = 0; i < NUM_REGS; i++) { reg[i] = 0; }
    for (int i = 0; i < USER_SZ; i++) { user[i] = 0; }
    // for (int i = 0; i < NUM_FUNCS; i++) { func[i] = 0; }
    HERE = &user[0];
}

void setCell(byte* to, CELL val) {
#ifdef _NEEDS_ALIGN_
    *(to++) = (byte)val; 
    for (int i = 1; i < CELL_SZ; i++) {
        val = (val >> 8);
        *(to++) = (byte)val;
    }
#else
    * ((CELL *)to) = val;
#endif
}

CELL getCell(byte* from) {
    CELL val = 0;
#ifdef _NEEDS_ALIGN_
    from += (CELL_SZ - 1);
    for (int i = 0; i < CELL_SZ; i++) {
        val = (val << 8) + *(from--);
    }
#else
    val = *((CELL*)from);
#endif
    return val;
}

void printStringF(const char* fmt, ...) {
    static char buf[128];
    va_list args;
    va_start(args, fmt);
    vsnprintf(buf, sizeof(buf), fmt, args);
    va_end(args);
    printString(buf);
}

void dumpStack() {
    printChar('(');
    for (int i = 1; i <= dsp; i++) {
        printStringF("%s%ld", (i > 1 ? " " : ""), (CELL)dstack[i].i);
    }
    printChar(')');
}

void skipTo(byte to, int isCreate) {
    while (*pc) {
        ir = *(pc++);
        if (isCreate) { *(HERE++) = ir; }
        if ((to == '"') && (ir != to)) { continue; }
        if ((to == '`') && (ir != to)) { continue; }
        if (ir == to) { return; }
        if (ir == '\'') { ++pc; continue; }
        if (ir == '(') { skipTo(')', isCreate); continue; }
        if (ir == '[') { skipTo(']', isCreate); continue; }
        if (ir == '"') { skipTo('"', isCreate); continue; }
        if (ir == '`') { skipTo('`', isCreate); continue; }
    }
    isError = 1;
}

void doFor() {
    CELL f = pop(), t=pop();
    lsp += 3; if (LSTACK_SZ < lsp) { printString("-[lsp]-"); lsp=LSTACK_SZ; }
    L0 = (f<t)?f:t; L1 = (t>f)?t:f; L2 = (CELL)pc;
}

int isOk(int exp, const char* msg) {
    isError = (exp == 0); if (isError) { printString(msg); }
    return (isError == 0);
}

void doFloat() {
    ir = *(pc++);
    switch (ir) {
        case  'F': FTOS = (double)TOS;
        RCASE 'I': TOS = (CELL)FTOS;
        RCASE '+': FNOS += FTOS; pop();
        RCASE '-': FNOS -= FTOS; pop();
        RCASE '*': FNOS *= FTOS; pop();
        RCASE '/': if (isOk(FTOS!=0, "-0div-")) { FNOS /= FTOS; pop(); }
        RCASE '<': FNOS = (FNOS<FTOS)  ? 1 : 0; pop();
        RCASE '>': FNOS = (FNOS>FTOS)  ? 1 : 0; pop();
        RCASE '=': FNOS = (FNOS==FTOS) ? 1 : 0; pop();
        RCASE '.': printStringF("%g", FTOS); pop();
        return; default:
            isError = 1;
            printStringF("-flt:%c?-", ir);
    }
}

int getRFnum(int max) {
    CELL n = 0;
    while (isRegChar(*pc)) { n = (n * 26) + *(pc++)-'A'; }
    if (BTWI(n,0,max-1)) { push(n); return 1; }
    printString("-RF#-");
    return 0;
}

void doExt() {
    ir = *(pc++);
    switch (ir) {
        case 'I': ir = *(pc++);
            if (ir == 'A') { 
                ir = *(pc++);
                if (ir == 'F') { push((CELL)&func[0]); }
                if (ir == 'H') { push((CELL)&HERE); }
                if (ir == 'R') { push((CELL)&reg[0]); }
                if (ir == 'U') { push((CELL)&user[0]); }
                if (ir == 'V') { push((CELL)&vars[0]); }
                return;
            };
            if (ir == 'C') { push(CELL_SZ); }
            if (ir == 'F') { push(NUM_FUNCS); }
            if (ir == 'H') { push((CELL)HERE); }
            if (ir == 'R') { push(NUM_REGS); }
            if (ir == 'U') { push(USER_SZ); }
            if (ir == 'V') { push(VARS_SZ); }
        RCASE 'K': dumpStack();
        RCASE 'S': if (*pc == 'R') { ++pc; vmInit(); }
        RCASE 'M': push(doMicros());
        RCASE 'T': push(doMillis());
        RCASE 'W': if (0 < TOS) { doDelay(TOS); } pop();
        RCASE 'R': if (!seed) { seed = getSeed(); }
                seed ^= (seed << 13);
                seed ^= (seed >> 17);
                seed ^= (seed << 5);
                TOS = (TOS) ? ABS(seed) % TOS : seed;
                return;
        default:
        pc = doCustom(ir, pc);
    }
}

addr run(addr start) {
    pc = start;
    isError = 0;
    rsp = lsp = locStart = 0;
next:
    if (isError) { return pc; }
    ir = *(pc++);
    switch (ir) {
        case 0:  return pc;
        NCASE ' ': while (*(pc) == ' ') { pc++; }
        NCASE '!': setCell((byte*)TOS, NOS); DROP2;
        NCASE '"': while (*(pc)!=ir) { printChar(*(pc++)); }; ++pc;
        NCASE '#': push(TOS);
        NCASE '$': t1 = TOS; TOS = NOS; NOS = t1;
        NCASE '%': push(NOS);
        // NCASE '&': /*FREE*/
        NCASE '\'': push(*(pc++));
        NCASE '(': if (!TOS) { skipTo(')', 0); } pop();
        NCASE ')': /* nothing to do here */
        NCASE '*': t1 = pop(); TOS *= t1;
        NCASE '+': t1 = pop(); TOS += t1;
        NCASE ',': printChar((char)pop());
        NCASE '-': t1 = pop(); TOS -= t1;
        NCASE '.': printStringF("%ld", (CELL)pop());
        NCASE '/': if (isOk(TOS, "-0div-")) { NOS /= TOS; pop(); }
        NCASE '0': case '1': case '2': case '3': case '4': case  '5': case '6':
        case '7': case '8': case '9': push(ir-'0');
            while (BTWI(*pc, '0', '9')) { TOS = (TOS * 10) + *(pc++) - '0'; }
            NCASE ':': if (!getRFnum(NUM_FUNCS)) { NEXT; }
            while (*(pc) == ' ') { pc++; }
            func[pop()] = (addr)pc;
            skipTo(';', 0);
            HERE = (HERE < pc) ? pc : HERE;
        NCASE ';': pc = rpop(); locStart -= (9<locStart) ? 10 : 0; if (!pc) return pc;
        NCASE '<': t1 = pop(); TOS = (TOS <  t1) ? 1 : 0;
        NCASE '=': t1 = pop(); TOS = (TOS == t1) ? 1 : 0;
        NCASE '>': t1 = pop(); TOS = (TOS >  t1) ? 1 : 0;
        // NCASE '?': /*FREE*/
        NCASE '@': TOS = getCell((byte*)TOS);
        NCASE 'A': if (TOS < 0) { TOS = -TOS; }
        NCASE 'B': printChar(' ');
        NCASE 'C': ir = *(pc++);
            if (ir == '@') { TOS = *(byte*)TOS; }
            else if (ir == '!') { *(byte*)TOS = (byte)NOS; DROP2; }
        NCASE 'D': --TOS;
        // NCASE 'E': /*FREE*/
        NCASE 'F': doFloat();
        NCASE 'G': if (TOS) { pc = (addr)TOS; } pop();
        // NCASE 'H': /*FREE*/
        NCASE 'I': push(L0);
        // NCASE 'J': /*FREE*/
        NCASE 'K': ir = *(pc++);
            if (ir == '?') { push(qkey()); }
            else if (ir == '@') { push(key()); }
        NCASE 'L': t1 = pop(); TOS = (TOS << t1);
        NCASE 'M': if (isOk(TOS, "-0div-")) { t1 = pop(); TOS %= t1; }
        NCASE 'N': printString("\r\n");
        // NCASE 'O': /*FREE*/
        NCASE 'P': ++TOS;
        // NCASE 'Q': /*FREE*/
        NCASE 'R': t1 = pop(); TOS = (TOS >> t1);
        NCASE 'S': if (isOk(TOS, "-0div-")) { t1 = TOS; TOS = NOS % t1; NOS /= t1; }
        // NCASE 'T': /*FREE*/
        NCASE 'U': TOS += (CELL)&user[0];
        NCASE 'V': TOS += (CELL)&vars[0];
        // NCASE 'W': /*FREE*/
        NCASE 'X': if (TOS) { rpush(pc); pc = (addr)TOS; } pop();
        // NCASE 'Y': /*FREE*/
        NCASE 'Z': printString((char*)pop());
        NCASE '[': doFor();
        NCASE '\\': pop();
        NCASE ']': if ((++L0)<L1) { pc=(addr)L2; NEXT; } /* fall through */
        case '^': lsp = 3; if (lsp<0) { lsp=0; }
        NCASE '_': TOS = (TOS) ? 0 : 1;
        NCASE '`': push(TOS);
            while ((*pc) && (*pc != ir)) { *(AOS++) = *(pc++); }
            *(AOS++) = 0; pc++;
        // NCASE 'a': /*FREE*/
        NCASE 'b': ir = *(pc++);                    // BIT operations
            if (ir == '&') { NOS &= TOS; pop(); }           // AND
            else if (ir == '|') { NOS |= TOS; pop(); }      // OR
            else if (ir == '^') { NOS ^= TOS; pop(); }      // XOR
            else if (ir == '~') { TOS = ~TOS; }             // NOT (COMPLEMENT)
            else if (ir == 'L') { blockLoad(pop()); }       // Block Load
            else if (ir == 'A') { loadAbort(); }            // Block Load Abort
            else if (ir == 'E') { doEditor(); }             // Block Edit
        NCASE 'c': if (getRFnum(NUM_FUNCS) && func[TOS]) {
            if (*pc != ';') { rpush(pc); locStart += 10; }
            pc = func[TOS];
        } pop();
        NCASE 'd': if (isLocal(*pc)) { --locals[*(pc++) - '0' + locStart]; }
                  else { if (getRFnum(NUM_REGS)) { --reg[pop()]; } }
        // NCASE 'e': /*FREE*/
        NCASE 'f': ir = *(pc++);
            if (ir == 'O') { fileOpen(); }
            else if (ir == 'C') { fileClose(); }
            else if (ir == 'R') { fileRead(); }
            else if (ir == 'W') { fileWrite(); }
            else if (ir == 'S') { codeSave(user, HERE); }
            else if (ir == 'L') { HERE = codeLoad(user, HERE); }
        // NCASE 'g': /*FREE*/
        NCASE 'h': push(0); while (1) {
                t1 = BTWI(*pc,'0','9') ? (*pc)-'0' : -1;
                t1 = BTWI(*pc,'A','F') ? (*pc)-'A'+10 : t1;
                if (t1 < 0) { NEXT; }
                TOS = (TOS * 16) + t1; ++pc;
            }
        NCASE 'i': if (isLocal(*pc)) { ++locals[*(pc++) - '0' + locStart]; }
                  else { if (getRFnum(NUM_REGS)) { ++reg[pop()]; } }
        // NCASE 'j': /*FREE*/
        // NCASE 'k': /*FREE*/
        // NCASE 'l': /*FREE*/
        // NCASE 'm': /*FREE*/
        // NCASE 'n': /*FREE*/
        // NCASE 'o': /*FREE*/
        NCASE 'p': L0 += pop();
        // NCASE 'q': /*FREE*/
        NCASE 'r': if (isLocal(*pc)) { push(locals[*(pc++)-'0'+locStart]); }
                  else { if (getRFnum(NUM_REGS)) { TOS = reg[TOS]; } }
        NCASE 's': if (isLocal(*pc)) { locals[*(pc++)-'0'+locStart] = pop(); }
                  else { if (getRFnum(NUM_REGS)) { reg[TOS] = NOS; DROP2; } }
        // NCASE 't': /*FREE*/
        // NCASE 'u': /*FREE*/
        // NCASE 'v': /*FREE*/
        // NCASE 'w': /*FREE*/
        NCASE 'x': doExt();
        // NCASE 'y':
        // NCASE 'z':
        NCASE '{': if (!TOS) { skipTo('}', 0); NEXT; }
                lsp += ((lsp+2) < LSTACK_SZ) ? 3 : 0;
                L0 = 0; L1 = 1; L2 = (CELL)pc;
        // NCASE '|':  /*FREE*/
        NCASE '}': if (TOS) { pc=(addr)L2; } else { pop(); lsp = (2<lsp) ? lsp-3 : 0; }
        NCASE '~': TOS = -TOS;
        NEXT;
        default: printStringF("-[ir:%d?]-", ir); NEXT;
    }
    return pc;
}
